File: src/csharp/VRE.Vridge.API.Client/Remotes/VridgeRemote.cs
using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.IO;
using System.Linq;
using System.Net.Sockets;
using System.Runtime.CompilerServices;
using System.Text;
using System.Threading.Tasks;
using NetMQ;
using ProtoBuf;
using VRE.Vridge.API.Client.Helpers;
using VRE.Vridge.API.Client.Messages.Control;
using VRE.Vridge.API.Client.Messages.Control.Responses;
using VRE.Vridge.API.Client.Messages.v3.Broadcast;
using VRE.Vridge.API.Client.Messages.v3.Discovery;
using VRE.Vridge.API.Client.Proxy;
using VRE.Vridge.API.Client.Proxy.Broadcasts;
using VRE.Vridge.API.Client.Proxy.Controller;
using VRE.Vridge.API.Client.Proxy.HeadTracking;
using VRE.Vridge.API.Client.Remotes.Beacons;

namespace VRE.Vridge.API.Client.Remotes
{
    /// <summary>
    /// Handles communication between all available VRidge remote endpoints.
    /// </summary>
    public class VridgeRemote : IDisposable
    {
        private static readonly DiscoveryClient Discovery;

        // User config
        private readonly int reconnectFrequencyMs;
        private readonly int timeoutThresholdMs;
        private readonly Capabilities capabilities;

        // Access objects
        private readonly APIClient api;        
        private ControllerRemote controller;
        private HeadRemote head;
        private BroadcastProxy broadcasts;

        // State
        private DateTime lastConnectionAttempt;
        private bool isDisposed = false;

        static VridgeRemote()
        {
            Discovery = new DiscoveryClient();
        }

        public VridgeRemote(
            string serverIp,
            string appName,
            Capabilities capabilities,
            int reconnectFrequencyMs = 2000,
            int timeoutThresholdMs = 100)
        {
            this.reconnectFrequencyMs = reconnectFrequencyMs;
            this.timeoutThresholdMs = timeoutThresholdMs;
            this.capabilities = capabilities;

            api = new APIClient(serverIp, appName);
        }

        public event EventHandler<HapticPulse> HapticPulse; 

        /// <summary>
        /// Returns list of currently running API servers in reachable networks.
        /// </summary>
        public static List<VridgeServerBeacon> ActiveVridgeServers => Discovery.ActiveVridgeServers;        

        /// <summary>
        /// Returns controller remote if connection seems valid. Will return null if connection is not in healthy state.
        /// Always check for null as a indicator of connection state.
        /// </summary>        
        public ControllerRemote Controller
        {
            get
            {
                if (isDisposed)
                {
                    throw new ObjectDisposedException("You already disposed this remote. Create new one.");
                }

                lock (this)
                {
                    if (controller != null && !controller.IsDisposed)
                    {
                        return controller;
                    }

                    if (lastConnectionAttempt.AddMilliseconds(reconnectFrequencyMs) < DateTime.Now)
                    {
                        return TrySettingUpConnection() ? controller : null;
                    }

                    return null;
                }                
            }
        }

        /// <summary>
        /// Returns head remote if connection seems valid. Will return null if connection is not in healthy state.
        /// Always check for null as a indicator of connection state.
        /// </summary>
        public HeadRemote Head
        {
            get
            {
                if (isDisposed)
                {
                    throw new ObjectDisposedException("You already disposed this remote. Create new one.");
                }

                lock (this)
                {
                    if (head != null && !head.IsDisposed)
                    {
                        return head;
                    }

                    if (lastConnectionAttempt.AddMilliseconds(reconnectFrequencyMs) < DateTime.Now)
                    {
                        return TrySettingUpConnection() ? head : null;
                    }

                    return null;
                }
            }
        }

        public APIStatus Status
        {
            get
            {
                lock (this)
                {
                    return api.GetStatus();
                }                
            }
        }

        public void Dispose()
        {
            lock (this)
            {
                if (isDisposed) return;

                isDisposed = true;
                HapticPulse = null;
                DisconnectAllEndpoints();
                Discovery.Dispose();
                
            }
        }

        /// <summary>
        /// Checks if current instance of VRidgeRemote can communicate
        /// with API without establishing permanent connection.
        /// </summary>        
        public VridgeRemoteConnectionStatus PreConnectionCheck(int timeoutMs)
        {
            lock (this)
            {
                try
                {
                    // First, check if we can even query status
                    var status = api.GetStatus(timeoutMs);
                    if (status == null)
                    {
                        return VridgeRemoteConnectionStatus.Unreachable;
                    }

                    // Then verify if headTracking is available if we want it
                    if (capabilities.HasFlag(Capabilities.HeadTracking))
                    {
                        var endpoint = status.Endpoints.FirstOrDefault(x => x.Name == EndpointNames.HeadTracking);
                        
                        if (endpoint == null)
                        {
                            // This shouldn't happen, ever
                            return VridgeRemoteConnectionStatus.UnexpectedError;
                        }

                        if (endpoint.Code == (int)ControlResponseCode.InUse)
                        {
                            return VridgeRemoteConnectionStatus.InUse;
                        }
                    }

                    // And do the same for controllers; this can be refactored (DRY)                    
                    if (capabilities.HasFlag(Capabilities.Controllers))
                    {
                        var endpoint = status.Endpoints.FirstOrDefault(x => x.Name == EndpointNames.Controller);
                        if (endpoint == null)
                        {
                            return VridgeRemoteConnectionStatus.UnexpectedError;
                        }

                        if (endpoint.Code == (int)ControlResponseCode.InUse)
                        {
                            return VridgeRemoteConnectionStatus.InUse;
                        }
                    }

                    return VridgeRemoteConnectionStatus.Okay;
                }
                catch (Exception)
                {
                    // Discard it and return unknown error. 
                    // API clients can use lower level access to read actual errors instead of VRidge remote which is fire-and-forget api.
                    return VridgeRemoteConnectionStatus.UnexpectedError;
                }

            }
        }

        // Sets up connection and returns true if all links are established.
        private bool TrySettingUpConnection()
        {
            lastConnectionAttempt = DateTime.Now;

            // Make sure API server is alive
            var status = api.GetStatus(timeoutThresholdMs);
            if (status == null) return false;

            // Reset current connections, if exist
            DisconnectAllEndpoints();

            try
            {                
                if (capabilities.HasFlag(Capabilities.HeadTracking))
                {
                    var endpointStatus = status.Endpoints.FirstOrDefault(x => x.Name == EndpointNames.HeadTracking);
                    if (endpointStatus == null || endpointStatus.Code != (int) ControlResponseCode.OK)
                    {
                        return false;
                    }

                    head = new HeadRemote(api.CreateProxy<HeadTrackingProxy>());
                }

                if (capabilities.HasFlag(Capabilities.Controllers))
                {
                    var endpointStatus = status.Endpoints.FirstOrDefault(x => x.Name == EndpointNames.Controller);
                    if (endpointStatus == null || endpointStatus.Code != (int) ControlResponseCode.OK)
                    {
                        return false;
                    }

                    controller = new ControllerRemote(api.CreateProxy<ControllerProxy>());                    
                }

                // Subscribe to haptic pulse broadcasts, if controller proxy is in use
                if (controller != null)
                {                    
                    broadcasts = api.CreateProxy<BroadcastProxy>();

                    // Forward the events to long-lived event so API user doesn't
                    // have to resubscribe on reconnect
                    broadcasts.HapticPulseReceived += (s, e) => HapticPulse?.Invoke(this, e);
                }

                return true;
            }
            catch (Exception x)
            {
                Logger.Error("Error during API connection: " + x);

                // Cleanup possibly connected endpoints
                DisconnectAllEndpoints();
                
            }

            return false;
        }

        private void DisconnectAllEndpoints()
        {
            controller?.Dispose();
            head?.Dispose();
            broadcasts?.Dispose();
            controller = null;
            head = null;
            broadcasts = null;
        }
    }
}

File: src/java/VRE.Vridge.API/VRE.Vridge.API.Client/src/main/java/com/riftcat/vridge/api/client/java/remotes/VridgeRemote.java
package com.riftcat.vridge.api.client.java.remotes;

import com.google.protobuf.InvalidProtocolBufferException;
import com.riftcat.vridge.api.client.java.APIClient;
import com.riftcat.vridge.api.client.java.Capabilities;
import com.riftcat.vridge.api.client.java.EndpointNames;
import com.riftcat.vridge.api.client.java.control.ControlResponseCode;
import com.riftcat.vridge.api.client.java.control.responses.APIStatus;
import com.riftcat.vridge.api.client.java.control.responses.EndpointStatus;
import com.riftcat.vridge.api.client.java.proto.Beacon;
import com.riftcat.vridge.api.client.java.proto.BeaconOrigin;
import com.riftcat.vridge.api.client.java.proto.HapticPulse;
import com.riftcat.vridge.api.client.java.proxy.BroadcastProxy;
import com.riftcat.vridge.api.client.java.proxy.ControllerProxy;
import com.riftcat.vridge.api.client.java.proxy.HeadTrackingProxy;
import com.riftcat.vridge.api.client.java.proxy.IBroadcastListener;
import com.riftcat.vridge.api.client.java.remotes.beacons.VridgeServerBeacon;
import com.riftcat.vridge.api.client.java.remotes.beacons.VridgeServerBeaconList;
import com.riftcat.vridge.api.client.java.utils.APILogger;

import org.zeromq.ZBeacon;

import java.net.InetAddress;
import java.util.ArrayList;
import java.util.EnumSet;
import java.util.List;

public class VridgeRemote {

    private static DiscoveryClient discovery;

    // User config
    private final int reconnectFrequencyMs;
    private final int timeoutThresholdMs;
    private final EnumSet<Capabilities> capabilities;
    private final String appName;

    // Access objects
    private APIClient api;
    private ControllerRemote controller;
    private HeadRemote head;
    private BroadcastProxy broadcasts;

    // State
    private long lastConnectionAttempt;
    private boolean isDisposed = false;
    private List<IBroadcastListener> eventListeners = new ArrayList<IBroadcastListener>();

    static{
        discovery = new DiscoveryClient();
    }

    public VridgeRemote(
            String serverIp,
            String appName,
            EnumSet<Capabilities> capabilities,
            int reconnectFrequencyMs,
            int timeoutThresholdMs)
    {
        this.reconnectFrequencyMs = reconnectFrequencyMs;
        this.timeoutThresholdMs = timeoutThresholdMs;
        this.capabilities = capabilities;
        this.appName = appName;
        api = new APIClient(serverIp, appName);
    }

    /**
     * Returns list of currently running API servers in reachable networks
     * along with their IP addresses and identification information.
     */
    public static List<VridgeServerBeacon> getVridgeServers(){
        return discovery.getFreshServers();
    }

    /** Adds subscriber that will receive haptic feedback events from VRidge.
     * Possibly more in the future. */
    public void addEventListener(IBroadcastListener listener){
        eventListeners.add(listener);
    }

    /** Removes event listener added by addEventListener. */
    public void removeEventListener(IBroadcastListener listener){
        eventListeners.remove(listener);
    }

    public synchronized APIStatus getStatus(){
        try {
            return api.GetStatus();
        } catch (Exception e) {
            return null;
        }
    }

    /**
     * Returns controller remote if connection seems valid. Will return null if connection is not in healthy state.
     * Always check for null as a indicator of connection state.
     */
    public ControllerRemote getController()
    {
        if (isDisposed){
            throw new IllegalStateException("You already disposed this remote. Create new one.");
        }

        synchronized (this)
        {
            if (controller != null && !controller.isDisposed())
            {
                return controller;
            }

            if (lastConnectionAttempt + reconnectFrequencyMs < System.currentTimeMillis())
            {
                return trySettingUpConnection() ? controller : null;
            }

            return null;
        }
    }

    /**
     * Returns head remote if connection seems valid. Will return null if connection is not in healthy state.
     * Always check for null as a indicator of connection state.
     */
    public HeadRemote getHead()
    {
        if (isDisposed){
            throw new IllegalStateException("You already disposed this remote. Create new one.");
        }

        synchronized (this)
        {
            if (head != null && !head.isDisposed())
            {
                return head;
            }

            if (lastConnectionAttempt + reconnectFrequencyMs < System.currentTimeMillis())
            {
                return trySettingUpConnection() ? head : null;
            }

            return null;
        }
    }

    /**  Checks if current instance of VRidgeRemote can communicate with API
     * without establishing permanent connection.
     * Will timeout after 1 second without response. */
    public VridgeRemoteConnectionStatus preCheck(){
        synchronized (this){
            try{
                // First, check if we can even query status
                APIStatus status = api.GetStatus();
                if (status == null){
                    return VridgeRemoteConnectionStatus.Unreachable;
                }

                // Then verify if headTracking is available if we want it
                if (capabilities.contains(Capabilities.HeadTracking)){
                    EndpointStatus endpoint = status.findEndpoint(EndpointNames.HeadTracking);

                    if (endpoint == null){
                        // This shouldn't happen, ever
                        return VridgeRemoteConnectionStatus.UnexpectedError;
                    }

                    if (endpoint.Code == ControlResponseCode.InUse){
                        return VridgeRemoteConnectionStatus.InUse;
                    }
                }

                // And do the same for controllers; this can be refactored (DRY)
                if (capabilities.contains(Capabilities.Controllers)){
                    EndpointStatus endpoint = status.findEndpoint(EndpointNames.Controller);
                    if (endpoint == null){
                        return VridgeRemoteConnectionStatus.UnexpectedError;
                    }

                    if (endpoint.Code == ControlResponseCode.InUse){
                        return VridgeRemoteConnectionStatus.InUse;
                    }
                }

                return VridgeRemoteConnectionStatus.Okay;
            }
            catch (Exception e){
                // Discard it and return unknown error.
                // API clients can use lower level access to read actual errors instead of VRidge remote which is fire-and-forget api.
                return VridgeRemoteConnectionStatus.UnexpectedError;
            }

        }
    }

    /** Clears all active connections. This object cannot be used after disposing */
    public void dispose(){
        synchronized (this){
            isDisposed = true;
            eventListeners.clear();
            disconnectAllEndpoints();
            discovery.dispose();
        }
    }

    /** Sets up connection and returns true if all links are established. */
    private boolean trySettingUpConnection()
    {
        lastConnectionAttempt = System.currentTimeMillis();

        // Make sure API server is alive
        APIStatus status;
        try{
            status = api.GetStatus();
        }
        catch (Exception x){
            return false;
        }

        // Reset current connections, if exist
        disconnectAllEndpoints();

        try
        {
            if (capabilities.contains(Capabilities.HeadTracking))
            {
                EndpointStatus endpointStatus = status.findEndpoint(EndpointNames.HeadTracking);
                if (endpointStatus == null || endpointStatus.Code != ControlResponseCode.OK)
                {
                    return false;
                }

                head = new HeadRemote((HeadTrackingProxy)api.getProxy(APIClient.HEADTRACKING));
            }

            if (capabilities.contains(Capabilities.Controllers))
            {
                EndpointStatus endpointStatus = status.findEndpoint(EndpointNames.Controller);
                if (endpointStatus == null){
                    return false;
                }

                controller = new ControllerRemote((ControllerProxy) api.getProxy(APIClient.CONTROLLER));
            }

            // Subscribe to haptic pulse broadcasts, if controller proxy is in use
            if (controller != null){
                broadcasts = api.getProxy(APIClient.BROADCASTS);

                // Forward the events to long-lived event so API user doesn't
                // have to resubscribe on reconnect
                broadcasts.addListener(new IBroadcastListener() {
                    @Override
                    public void onHapticPulse(HapticPulse pulse) {
                        for (IBroadcastListener listener : eventListeners) {
                            listener.onHapticPulse(pulse);
                        }
                    }
                });

                broadcasts.startPolling();
            }

            return true;
        }
        catch (Exception x)
        {
            APILogger.error("Error during API connection: " + x);

            // Cleanup possibly connected endpoints
            if(controller != null) controller.dispose();
            if(head != null) head.dispose();
            controller = null;
            head = null;
        }

        return false;
    }

    private void disconnectAllEndpoints(){
        if(head != null) head.dispose();
        if(controller != null) controller.dispose();
        if(broadcasts != null) broadcasts.disconnect();
        head = null;
        controller = null;
    }
}
